### ISSUE TITLE
Map-ol: support WFS Layers.

### ISSUE THREAD

**Issue by MichaelLangbein** [dlr-eoc/ukis-frontend-libraries#73]
<img src="https://www.eclipse.org/community/eclipse_newsletter/2014/march/images/article1.3.png" width=400 height=300>

## Description
While there is a LayerType 'wfs' in `services-layers`, currently when adding a new `VectorLayer` of type 'wfs', nothing is being displayed in a ol-map. The reason is that `map-ol` does not currently support the conversion of UkisLayers of type 'wfs' to OlLayers.

We're [currently in the process of adding support](https://github.com/dlr-eoc/ukis-frontend-libraries/tree/prereleaseWuekisSprint), but there are a few design decisions that I'd like to discuss.

#### CQL-filters
Occasionally, we want a user to be able to specify filters with their request. The question is, then, where we should specify those filters?

Naively, one option would be to add a new field to `services-layers/IVectorLayerOptions` like so:
```typescript
export interface IVectorLayerOptions extends ILayerOptions {
  /** data: geojson data */
  data?: any;
  url?: string;
  subdomains?: Array<string>;
  /** vector options like style, pointToLayer... depends on the map-library */
  options?: {
    /** ol/style/Style */
    style: any;
    [k: string]: any;
  };
  /** Only if type is 'wfs'. Allows user to only select certain properties of a  WFS-feature */
  cqlFilter?: string;
  /** if true clusters points | or set a Object with cluster options e.g. distance ... depends on the map-library */
  cluster?: boolean | IAnyObject;
  type: TVectorLayertype;
}
```

But there may be a better place for this - I'm very happy for all feedback on this!

#### Dimensions
`services-layers/ILayerOptions` already has a field `dimensions`. 
In our timeslider this field is used for WMS-Layers to create dots where time-values are.
The WFS protocol, contrary to WMS, [does not have a `TIME` parameter](https://stackoverflow.com/questions/47274529/wms-and-wfs-time-search-criterion-dimension). Yet the case can come up where a WFS-feature has different properties that are associated with different timesteps. 

Maybe we should reflect this case in the `dimensions` property. For example, the interface `ILayerTimeDimension` might contain a field mapping feature-properties to timestamps. 

Again, I'm not biased here - this idea has pro's and con's.
 - Pro: allows to time-enable WFS'es just like WMS'es, so we can use them in the timeslider.
 - Con: Quite complicated. Mostly it bugs me that this might be [non-orthagonal](https://en.wikipedia.org/wiki/Orthogonality_(programming)#:~:text=Orthogonality%20in%20a%20programming%20language,the%20design%2C%20the%20fewer%20exceptions.) to the `CqlFilter` property discussed above.

* Does this change impact existing behaviors? If so how?
This would only extend `map-ol` functionality to WFS'es, not impact any existing features.

## Relevant Package
This feature request is for [@dlr-eoc/map-ol](https://github.com/dlr-eoc/ukis-frontend-libraries/packages)

## Examples
Are there any examples in other projects that demonstrate this feature or enhancement?

This issue arose from the currently ongoing work in WUEKIS, but also on CoastalX, both of which use a timeslider.

### Describe alternatives you've considered
Have you considered any alternative solutions or workarounds?

*Yes*, the obvious alternative to adding WFS support to map-ol is using `CustomLayer`s.

Before, when using WFS'es in openlayers maps, we have created instances of `CustomLayer`. However, `CustomLayer` is designed to allow the programmer to manually create a ol-layer. That is quite specific.

In many of our projects, we read layers, including WFS'es, from configuration files (like OWC-json). We might want to use those UkisLayers in cesium or threejs-maps as well, so creating a CustomLayer here is not really a sustainable solution.

In other words, this works easily:
``` owc-json -> UkisVectorLayer -> any mapping library```
Whereas this requires custom work for each mapping library:
``` 
owc-json -> CustomLayer<ol> -> map-ol
owc-json -> CustomLayer<three> -> map-three
owc-json -> CustomLayer<cesium> -> map-cesium
```

**Comment by boeckMt** [comment]
### WFS and CQL

I think we only used the `VectorLayer`.[url](https://github.com/dlr-eoc/ukis-frontend-libraries/blob/v7.3.0-next.2/projects/services-layers/src/lib/types/Layers.ts#L318) for this till now.

E.g. in other Projects we used `IOwsOffering` and converted the data then with [OwcJsonService](https://github.com/dlr-eoc/ukis-frontend-libraries/blob/v7.3.0-next.2/projects/services-ogc/src/lib/owc/owc-json.service.ts) ([createVectorLayerFromOffering](https://github.com/dlr-eoc/ukis-frontend-libraries/blob/v7.3.0-next.2/projects/services-ogc/src/lib/owc/owc-json.service.ts#L351)) to a `VectorLayer` which can then be handled by `map-ol`.

```ts
const wfsLayer: IOwsOffering = {
    code: 'http://www.opengis.net/spec/owc-geojson/1.0/req/wfs',
    operations: [
        {
            code: 'GetFeature',
            method: 'GET',
            type: 'application/json',
            href: geoserverUrl + 'ows?service=WFS' +
                '&version=1.0.0' +
                '&request=GetFeature' +
                '&outputFormat=application/json' +
                '&typeName=' + Name + cqlFilter
        }
    ],
    title: layerTitle,
    reflectUrl: geoserverUrl + 'wms/reflect?layers=' + Name,
    legendUrl: geoserverUrl + 'ows?SERVICE=WMS' +
        '&VERSION=1.3.0' +
        '&REQUEST=GetLegendGraphic' +
        '&transparent=true' +
        '&FORMAT=image/png' +
        '&LAYER=' + lName,
    iconUrl: ''
}
```

--- 
Also we made a Example to use `WFS` here https://github.com/dlr-eoc/ukis-frontend-libraries/blob/b52f5c5e6ed38d6e70ecbf3e50e1436cdb2ad0b6/projects/demo-maps/src/app/route-components/route-example-olperformance/services/largelayers.service.ts#L123


---

For the Interface we could create a new Type, like it was done with [WmtsLayer](https://github.com/dlr-eoc/ukis-frontend-libraries/blob/v7.3.0-next.2/projects/services-layers/src/lib/types/RasterLayers.ts#L19) then we can add 

```ts
...
/** Only if type is 'wfs'. Allows user to only select certain properties of a  WFS-feature */
  cqlFilter?: string;
...
```

but this is than like an update function for the Layer.url, so you need to use getter and setter to update the url when the `cqlFilter` is changing. I don't know if it is not better to only use the url for this, so you can customly create the url for the request you like.

--- 

### Dimensions

The Layer.dimensions object is already a bit complicated...

It allows `time` and `elevation`. In the time object you can set [values](https://github.com/dlr-eoc/ukis-frontend-libraries/blob/v7.3.0-next.2/projects/services-layers/src/lib/types/Layers.ts#L145) which are time steps or Intervals.

Then for time we also have [Layer.time](https://github.com/dlr-eoc/ukis-frontend-libraries/blob/v7.3.0-next.2/projects/services-layers/src/lib/types/Layers.ts#L107) which can be a string for the current timestep to use. 

I think we should try to make all the time stuff easier and not more complicated like it is now :)

**Comment by boeckMt** [comment]
I also found this GeoSolutions post about [Vector Data dimensions](https://geoserver.geo-solutions.it/educational/en/multidim/vector/index.html)

To request dimensions like time, we have to use the WFS filters:
- https://www.ogc.org/standards/filter 
- https://docs.geoserver.org/latest/en/user/filter/filter_reference.html

Here are also two examples from Openlayers to use WFS
- https://openlayers.org/en/latest/examples/vector-wfs.html
- https://openlayers.org/en/latest/examples/vector-wfs-getfeature.html

**Comment by MichaelLangbein** [comment]
Roger that!

> I think we only used the VectorLayer.url for this till now.

Very well - I've updated the code accordingly, and created an example in ´demo-maps´. This certainly makes the code a lot simpler. If there's a need for it, we can still go ahead and flesh out a potential class `WfsLayer` in the future, but for now, this is certainly the most straightforward solution.

> I think we should try to make all the time stuff easier and not more complicated like it is now :)

Completely agree! As for now, there is no interaction between dimension and any methods on layers.


All in all, this is a [pretty minimal PR](https://github.com/dlr-eoc/ukis-frontend-libraries/pull/76). I hope this is on one hand useful and on the other hand simple enough to be potentially expanded on in the future. 

Thanks a lot!

**Comment by MichaelLangbein** [comment]
Closing this issue now that the PR is done. Feel free to re-open or create a new one for more WFS-specific functionality.
